### PR TITLE
bumpfee: Allow the user to choose which output is change

### LINKED ISSUE #11233: bumpfee behavior with custom change address
If I create an RBF transaction using the Qt GUI but specify my own custom change address, bumpfee doesn't work because it claims it can't find the change address:

    "Transaction does not have a change output"

This is because none of the outputs satisfy

    pWallet->IsChange(...)

I see there's a "TODO" commented in `CWallet::IsChange()` in `src/wallet/wallet.cpp`:

    // TODO: fix handling of 'change' outputs. The assumption is that any
    // payment to a script that is ours, but is not in the address book
    // is change.

I tend to put all my addresses in the address book whether I'm going to use them as change or not. I guess if the wallet was capable of creating its own SegWit change addresses I'd stop using custom change addresses, but currently it always seems to send change to a legacy address if I don't manually specify a change address.

### LINKED ISSUE #20795: Privacy Issue - Increase Fee with Custom Change Address grabs new UTXO
I am relaying a privacy issue for Bitcoin Core that was [reported on r/Bitcoin](https://reddit.com/r/Bitcoin/comments/klk2wh) (Reddit).

The [message](https://pastebin.com/raw/ptFQityt) was:

- _Using coin control_
- _Custom change address_

_When one clicks the increase transaction fee button, it adds another input rather than increasing the fee by reducing the change output. Major privacy leak since you have no control of the input used._

From that it sounds like the observed behavior is that the replacement transaction includes an additional UTXO that was not chosen by the user, even though the transaction was composed as a Coin Control transaction.

So I'm assuming the expected behavior would be for Bitcoin Core to remember that this OUTPUT was actually a Custom Change Address and to then lower the amount of that OUTPUT by the amount needed for the fee bump.

And, now that I am thinking of this, isn't there the expectation that if my initial transaction was done using Coin Control, and then if my fee bump (RBF) needs another UTXO input to cover the increase in fee, that Coin Selection for that will occur (regardless of whether or not the change address was a Custom Change Address)?

### PR DESCRIPTION
**Pull request by achow101**
When bumping the transaction fee, we will try to find the change output of the transaction in order to have an output whose value we can modify so that we can meet the target feerate. However we do not always find the change output which can cause us to unnecessarily add an additional output to the transaction. We can avoid this issue by outsourcing the determination of change to the user if they so desire.

This PR adds a `orig_change_pos` option to bumpfee which the user can use to specify the index of the change output.

Fixes #11233
Fixes #20795

### CODE REVIEW

**Review by luke-jr — CHANGES_REQUESTED** [review]
We support subtracting fee from output, so I think the `"reduce_output"` name used in #12096 was better. (Would be nice if that situation could be detected, but I don't think we save the metadata needed for that.)

Also, might make sense to split the GUI part off to the GUI repo after the rest is merged. Seems like a hard thing to get a good UX for.

**Comment by 1440000bytes** [comment]
Concept ACK

**Comment by john-moffett** [comment]
Concept ACK. 

Tested and code reviewed 73d4ba2. Suggest changing the dialog UI to support scrolling and resizing in case there are lots of outputs:

<details><summary>Screenshot</summary>
image
</details>

[Here is a drop-in gist](https://gist.github.com/john-moffett/87c0cf77ed9827e05c92066254335c0a) to implement scrolling and resizing, which changes it to:

<details><summary>Screenshot</summary>
image
</details>

If a user creates a single-output tx and chooses that output as "change" to fee-bump, it will fail with the "Transaction must have at least one recipient" message. This makes sense if the user always expects the "payment" output to remain the same value, but if they're consolidating UTXOs or something, this may be desirable to allow. 

Finally, I suggest having the "None" radio box be selected by default (if no change outputs are found). The code acts that way if nothing gets selected, but it could be confusing if the user clicks "OK" with no radio boxes checked.

**Comment by Sjors** [comment]
> We support subtracting fee from output, so I think the `"reduce_output"` name used in #12096 was better.

Agreed. The help text could clarify that in the general case you should pick the change address output for this.

**Comment by achow101** [comment]
> We support subtracting fee from output, so I think the `"reduce_output"` name used in #12096 was better. (Would be nice if that situation could be detected, but I don't think we save the metadata needed for that.)

I've taken the name, but I'm not entirely sure that it is accurate. While we will most likely reduce the amount of the specified output, it's not the only outcome. It could be that specified output does not provide enough value, in which case we will add additional inputs and set the output's value as the resulting change value, which will be higher than the original value.

> Also, might make sense to split the GUI part off to the GUI repo after the rest is merged. Seems like a hard thing to get a good UX for.

Done, moved the GUI commits to https://github.com/bitcoin-core/gui/pull/700

> Suggest changing the dialog UI to support scrolling and resizing in case there are lots of outputs:

Took this suggestion in the PR to the GUI repo.

> If a user creates a single-output tx and chooses that output as "change" to fee-bump, it will fail with the "Transaction must have at least one recipient" message. This makes sense if the user always expects the "payment" output to remain the same value, but if they're consolidating UTXOs or something, this may be desirable to allow.

Added a commit which handles that case.

**Comment by Sjors** [comment]
Concept ACK

Did you mean for the middle commit to have `txid:` as its prefix?

**Comment by achow101** [comment]
> Did you mean for the middle commit to have `txid:` as its prefix?

Nope, fixed.

> On signet with an empty mempool, when I send a transaction with a single input at 1 sat/vbyte and then try to bump it to 2 sat/byte, it complains:
> 
> ```
> Insufficient total fee 0.00000222, must be at least 0.00000666 (oldFee 0.00000111 + incrementalFee 0.00000555)
> ```
> 
> `getnetworkinfo` says `"incrementalfee": 0.00001000`
> 
> Maybe `maxTxSize` is too pessimistic?

I'm going to say that this is an orthogonal issue. It looks like the wallet has a minimum incremental feerate of 5 sat/vb, rather than the 1 sat/vb the mempool uses.

File: src/wallet/feebumper.cpp
Comment (Sjors) [inline src/wallet/feebumper.cpp:300]: 0b280affd4c755d8f9eb809b3676389b2c2c52c0 nit: `return` could be outside the `if else`. If you touch it, there's also a typo in the commit description: `transcation`

**Review by Sjors — APPROVED** [review]
tACK 0b280affd4c755d8f9eb809b3676389b2c2c52c0

**Comment by Sjors** [comment]
Looks like this has a silent conflict with master for `wallet_bumpfee.py`.

**Comment by Sjors** [comment]
utACK 4167843e0729994954317e2a4ac8a96a453bad79. The latest rebase takes the new `outputs` argument into account from #25344. It ensures they can't be combined, so that's good.

File: test/functional/wallet_bumpfee.py
Comment (pinheadmz) [inline test/functional/wallet_bumpfee.py:205]: Will `new_change_pos == change_pos` always here? Or are the outputs of the replacement re-randomized?

**Review by pinheadmz — APPROVED** [review]
ACK 4167843e0729994954317e2a4ac8a96a453bad79

reviewed code, ran tests. Played with the feature on regtest and in particular using `reduce_output` to intentionally reduce the recipient output instead of the original change output added automatically by `sendtoaddress`. I don't think is explicitly covered in the test but could be. One other quesiton about the test below but otherwise LGTM

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 4167843e0729994954317e2a4ac8a96a453bad79
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmP4544ACgkQ5+KYS2KJ
yTr/hw//RCFzObS/NFeiJjG4Btb6kuDMDW7SYmXaEazT0mX91QpiQlq+cW92/FZa
FDqycJs43hMMu0ytmTJ8qs9S0sROU0IYRJghK16Ne16TBhtJwbpS93NIQo4LQY7j
a77Ql9DmDGt/c5EkN+lQsvk6W6PXlo1NqwvXT5Rn3NXnbJ0yD4PCo5ZQW1GQ305i
6pGboMlx97PGRtnCsFvfO6+lOAUlAPiwYXt8a+yfT2jhhFzmoi0KG4NlcPLHmwCL
OPKken36AE/2WMpZLu12nDMAb6qM3ia7/GiKxEB6+C0j0RBxYF4RAjmKxgNo/XJ/
f2yZiPmDSsGnZWaYRMpOjR5s0HgQS98WRUK17ZJfz8tN7qR9Y8L8wUwLTpvqHM/a
i+Je0mXY1k5iUtkOG9D9PKAAmsQ6OQzoCW1vzuKL1Z1jc2nKZHbuYbuE8OzyKlGc
A4gv/pNSyaWLHP7TqwvzxHTVlCNlbp+bN+PB21ojd4xsItCU5AGbLAxAMF+LJnwO
YL5xI+29voZcnZBiQI9/jHV8VJ5mUYYDT4f/nQYkTlMmAcBx7BzPF4Zh3h7gKOBA
ybhrHHsSNiXVcm2VIqI6Zb/Wvn/zRTJscxvRQNDuYvEGeCMjQSc1uaGulFflWLmq
qtGtR6U4j4GRTL8TyQcAZu4VfuaUH3up2Km3cKAAFvNqbaKOqlo=
=A42I
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

File: src/wallet/feebumper.cpp
Comment (furszy) [inline src/wallet/feebumper.cpp:252]: In https://github.com/bitcoin/bitcoin/commit/a902eaf5be4a83dff37f28a988e5f3b2425551c0:

Could write it prettier:
```suggestion
        if (reduce_output.has_value() ? reduce_output.value() == i : OutputIsChange(wallet, output)) {
```

**Review by furszy — COMMENTED** [review]
Concept ACK, left two nits. Will keep reviewing.

File: src/wallet/feebumper.cpp
Comment (furszy) [inline src/wallet/feebumper.cpp:324]: This is not contemplating the user changing the outputs for a single change output. E.g. bump tx to send coins back to yourself.

The generated `temp_mtx` uses the same outputs as the previous transaction, so the provided output is discarded. And only deduces the new fee from the previous tx outputs.

Previously, in master, we would just fail on this situation. While here, we return a tx that is not what the user wants.
So, created #27195 to solve the issue in master first.

File: test/functional/wallet_bumpfee.py
Comment (achow101) [reply test/functional/wallet_bumpfee.py:205]: They will be re-randomized.

File: src/wallet/feebumper.h
Comment (achow101) [reply src/wallet/feebumper.h:46]: Done

File: src/wallet/feebumper.cpp
Comment (achow101) [reply src/wallet/feebumper.cpp:252]: Done

File: src/wallet/feebumper.cpp
Comment (achow101) [reply src/wallet/feebumper.cpp]: I've added a fix, and adopted your test. The fix is simple enough to do here that I think a separate PR is unnecessary.

File: src/wallet/feebumper.cpp
Comment (furszy) [inline src/wallet/feebumper.cpp:319]: There still is a problem with this approach when the user sets a value for the single output.

Basically, the user provided value is being discarded. We don't check if it's within the inputs amount bounds nor if need to create a change output for the remaining amount (value can be lower than the inputs amount).

Current code instead of create a change output when the value is lower than the inputs amount, it sets the value equal to inputs total minus new fee. We should code a test for it too.

Thus why thought about using SFFO in 27195 to mark the output from where we want to subtract fees, so the apparent "single output tx" can be crafted using `CreateTransaction` instead of making an special case here.

Side from that, all good about closing 27195. I made it just because it's a bug in master.

Comment (achow101) [reply src/wallet/feebumper.cpp:319]: Ah, I see. Actually I've decided to just rebase this PR on top of #27195.

File: src/wallet/feebumper.cpp
Comment (achow101) [reply src/wallet/feebumper.cpp:252]: Fixed.

File: src/wallet/feebumper.cpp
Comment (furszy) [reply src/wallet/feebumper.cpp]: k, either way is fine for me. Cool about the special case exception removal and the new test.
just updated 27195 with further tests too.

**Comment by RandyMcMillan** [comment]
Concept ACK

File: src/wallet/feebumper.cpp
Comment (ismaelsadeeq) [inline src/wallet/feebumper.cpp:158]: Is this not redundant since already checked in spend.cpp? before calling `CreateRateBumpTransaction`?

**Comment by ismaelsadeeq** [comment]
tACK dff7235f9a6
Tested the `reduce_output` option on `regtest` with `bumpfee` RPC, the value of the output in the `reduced_output`  was decreased to fee bump the transaction instead of adding another input.
tests fail on master and pass on dff7235f9a6.
Code looks good to me

File: src/wallet/feebumper.cpp
Comment (ismaelsadeeq) [reply src/wallet/feebumper.cpp:158]: Could we perhaps consider relying on the check here? It seems like a duplicate of the same process in two different places if you agree.

Comment (achow101) [reply src/wallet/feebumper.cpp:158]: Done

**Comment by pinheadmz** [comment]
> Additionally, a new dialog is added to the GUI that allows the user to select the change output when they are bumping the transaction fee. This dialog will default select output that the wallet determines to be change, if it could find one.

Not seeing this in this branch.

**Review by pinheadmz — APPROVED** [review]
re-ACK e8c31f135c6e9a5f57325dbf4feceafd384f7762

Reviewed all code changes, ran tests locally. I played with the feature in regtest with wallet TXs as well as PSBTs. Made sure invalid input threw errors and expected output was reduced by fee bump. Found one nit below. Maybe remove GUI from OP description? Is that being added in a bitcoin-core/gui PR ?

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK e8c31f135c6e9a5f57325dbf4feceafd384f7762
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmS1V/kACgkQ5+KYS2KJ
yTon4g/+JpVTIdCwb4fvLOiulKLu+poVhET+Wz9iyC+47m4GJUPgcR8RyEcPFa+6
eEn7V1md7o9pOekPQnBLStZJZSFQwuJgG0utgt/M0a5yzQlNNsTcEh2nWqSAI1X2
RAzxOlBjomFzQndm8zQ6QVQRX4tmrwKdnJxumZxWrnRBToJCvikdDVzHIP1yS0CE
d9Z2yWNgJKkocuWACMOgfjt5GjR/Y0Y4qPDahcI10qpWjD2yFJMrx5DUjUnMxXyD
H/3+kYUga9yfFAlE1J9OyCWpL1k//zRcttBZxDoHZ/Hh3MIMvcla6eRAjndd0iU0
o92BVZS1+Ylz4mY2+aoja1yoXQa9JJwJavyOVNR0HtzXXtkzkWdf8shMcUe7jTXb
wlUJwm4+ItIZ05ms05eTmcxfDOc6aQjKud26hIG0TaztfLAZU3ITDdU9lHeZB8qC
szHpyVxLJ63BtaQCOdC//asAqxsn44EZlM3LY9iUE1jvxLAaaHN7yPgSO62A1B3N
RbQ+5BYLYWv6ceAnyUKxX7XnzEGjsFQnSe45KVYZ42zy4eO6kuc/xKy7his3oJ9r
7JT1FgA41wY8mi4KazB6JRyTwhcx0QQ3UxwbnLy3tzsTzWw0aU3dPqwG7iR2dTqH
Ac08JiqHzUCtTiYbRVrrATkgoZNeBqjCRhVr3o0z+2iKZLm6lWg=
=8/+K
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

**Comment by achow101** [comment]
> > Additionally, a new dialog is added to the GUI that allows the user to select the change output when they are bumping the transaction fee. This dialog will default select output that the wallet determines to be change, if it could find one.
> 
> Not seeing this in this branch.

That was moved to https://github.com/bitcoin-core/gui/pull/700. Dropped that from the PR description.

File: src/wallet/feebumper.cpp
Comment (pinheadmz) [reply src/wallet/feebumper.cpp:254]: Weird, this comment was posted yesterday but is dated over a month ago (Jun 9). I thought for sure I had tested this case locally and didn't see a missing output -- is this still the case here? Since we check `reduce_output.has_value()` in the if condition, doesn't that mean that if that option is set only the selected option will be designated as change?

Comment (furszy) [reply src/wallet/feebumper.cpp:254]: > I thought for sure I had tested this case locally and didn't see a missing output -- is this still the case here? Since we check `reduce_output.has_value()` in the if condition, doesn't that mean that if that option is set only the selected option will be designated as change?

Yeah. Seems that I was partially blind yesterday. This issue can still be triggered by a transaction with +2 change outputs and no `reduce_output` (`OutputIsChange()` will be true for them, which overwrites `destChange`). But.. it isn't related to this PR.
So.. nvm. Will do another review round. Thanks @pinheadmz.

**Review by furszy — COMMENTED** [review]
Code review ACK e8c31f13

File: src/wallet/rpc/spend.cpp
Comment (furszy) [reply src/wallet/rpc/spend.cpp:1022]: @whitslack yes. That would be an ugly outcome. We could have fixed the RPC docs here.

Would say to work towards #27601 to fix this scenario (and others) properly.
This issue comes from a workaround code that avoids the dup change outputs bug. Basically, the change output is manually discarded from the recipients list and expected to be re-added later, with the new fees subtracted, by the inner transaction creation process logic.

With #27601, we will be able to specify the output to reduce (SFFO) and the existent change output unequivocally.

Comment (luke-jr) [reply src/wallet/rpc/spend.cpp:1022]: This seems like a very serious bug that should have blocked the merge... :|

Comment (Sjors) [reply src/wallet/rpc/spend.cpp:1022]: Should we disable the adding of new inputs when `reduce_output` is set? If someone really intends to do that, they should probably use `outputs`.

Comment (Sjors) [reply src/wallet/rpc/spend.cpp:1022]: Made a separate issue to track this: #28180

**Review by donsheddy4 — COMMENTED** [review]
Very good

**Pull request merged**
